File: remote/src/main/java/com/kelsos/mbrc/interactors/LibraryAlbumInteractorImpl.java
package com.kelsos.mbrc.interactors;

import com.google.inject.Inject;
import com.kelsos.mbrc.domain.Album;
import com.kelsos.mbrc.mappers.AlbumMapper;
import com.kelsos.mbrc.repository.library.AlbumRepository;
import java.util.List;
import rx.Observable;
import rx.schedulers.Schedulers;

public class LibraryAlbumInteractorImpl implements LibraryAlbumInteractor {
  @Inject private AlbumRepository repository;

  @Override public Observable<List<Album>> execute(int offset, int limit) {
    return repository.getAlbumViews(offset, limit)
        .flatMap(albums -> Observable.just(AlbumMapper.map(albums)))
        .subscribeOn(Schedulers.io());
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/interactors/LibraryAlbumInteractorImpl.kt
package com.kelsos.mbrc.interactors

import com.kelsos.mbrc.domain.Album
import com.kelsos.mbrc.extensions.io
import com.kelsos.mbrc.mappers.AlbumMapper
import com.kelsos.mbrc.repository.library.AlbumRepository
import rx.Observable
import rx.lang.kotlin.toSingletonObservable
import javax.inject.Inject

class LibraryAlbumInteractorImpl
@Inject constructor(private val repository: AlbumRepository): LibraryAlbumInteractor {

  override fun getPage(offset: Int, limit: Int): Observable<List<Album>> {
    return repository.getAlbumViews(offset, limit)
        .flatMap { AlbumMapper.map(it).toSingletonObservable() }
        .io()
  }
}
